framework/aster-frame/src/mm/page_table/frame.rs

```diff
@@ -0,0 +1,515 @@
+//! This is implemented by using a reference counter in the frame metadata. If the above
+//! conditions are not met, the page table frame is ensured to be freed upon dropping the last
+//! reference.
+//!
+//! One can acquire exclusive access to a page table frame using merely the physical address of
+//! the page table frame. This is implemented by a lock in the frame metadata. Here the
+//! exclusiveness is only ensured for kernel code, and the processor's MMU is able to access the
+//! page table frame while a lock is held. So the modification to the PTEs should be done after
+//! the initialization of the entity that the PTE points to. This is taken care in this module.
+//!
+
+use core::{marker::PhantomData, mem::ManuallyDrop, ops::Range, panic, sync::atomic::Ordering};
+
+use super::{nr_subpage_per_huge, page_size, PageTableEntryTrait};
+use crate::{
+    arch::mm::{PageTableEntry, PagingConsts},
+    mm::{
+        paddr_to_vaddr,
+        page::{
+            allocator::FRAME_ALLOCATOR,
+            meta::{FrameMeta, PageMeta, PageTablePageMeta, PageUsage},
+            Page,
+        },
+        page_prop::PageProperty,
+        Frame, Paddr, PagingConstsTrait, PagingLevel, PAGE_SIZE,
+    },
+};
+
+/// The raw handle to a page table frame.
+///
+/// This handle is a referencer of a page table frame. Thus creating and dropping it will affect
+/// the reference count of the page table frame. If dropped the raw handle as the last reference,
+/// the page table frame and subsequent children will be freed.
+///
+/// Only the CPU or a PTE can access a page table frame using a raw handle. To access the page
+/// table frame from the kernel code, use the handle [`PageTableNode`].
+#[derive(Debug)]
+pub(super) struct RawPageTableNode<E: PageTableEntryTrait, C: PagingConstsTrait>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    pub(super) raw: Paddr,
+    pub(super) level: PagingLevel,
+    _phantom: PhantomData<(E, C)>,
+}
+
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> RawPageTableNode<E, C>
+where
+    [(); C::NR_LEVELS as usize]:,
+{
+    pub(super) fn paddr(&self) -> Paddr {
+        self.raw
+    }
+
+    /// Convert a raw handle to an accessible handle by pertaining the lock.
+    pub(super) fn lock(self) -> PageTableNode<E, C> {
+        let page = unsafe { Page::<PageTablePageMeta<E, C>>::restore(self.paddr()) };
+        debug_assert!(page.meta().level == self.level);
+        // Acquire the lock.
+        while page
+            .meta()
+            .lock
+            .compare_exchange(0, 1, Ordering::Acquire, Ordering::Relaxed)
+            .is_err()
+        {
+            core::hint::spin_loop();
+        }
+        // Prevent dropping the handle.
+        let _ = ManuallyDrop::new(self);
+        PageTableNode::<E, C> { page }
+    }
+
+    /// Create a copy of the handle.
+    pub(super) fn copy_handle(&self) -> Self {
+        core::mem::forget(unsafe { Page::<PageTablePageMeta<E, C>>::clone_restore(&self.paddr()) });
```

Comment:
As a rule, every usage of `unsafe` should be explained. There are over 30 locations of `unsafe`, many of them lack of a safety comment to justify its safety. This rule should be upheld despite of how clever you are or how confidence you are with the code because the rule helps reduce the odds of mistakes and improve code quality.

For example, if you have complied with this rule, then it would have been a pain in the ass for you to do the same reasoning for the following code pattern repeatedly.

```
let page = unsafe { Page::<PageTablePageMeta<E, C>>::restore(self.paddr()) };
```

The unsafe block has been repeated for four times. It should have been wrapped in a safe method so that you only need to write the safety reason once.

Reply:
> It should have been wrapped in a safe method

Sadly no. If wrapped the wrapping function would still be unsafe because the safety holds if the caller remembers to forget it properly. I'll add safety comments for all 4 cases.